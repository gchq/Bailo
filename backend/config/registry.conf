version: 0.1
log:
  fields:
    service: registry
storage:
  cache:
    blobdescriptor: inmemory
  s3:
    regionendpoint: http://minio:9000
    bucket: registry
    secure: false
    region: ignored-for-minio
    forcepathstyle: true
  redirect:
    disable: true
  delete:
    enabled: true
  maintenance:
    uploadpurging:
      enabled: true
      age: 168h
      interval: 24h
      dryrun: false
    readonly:
      enabled: false
http:
  addr: :5000
  relativeurls: true
  headers:
    X-Content-Type-Options: [nosniff]