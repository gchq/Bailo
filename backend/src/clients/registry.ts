import { Readable } from 'node:stream'

import { BodyInit, HeadersInit, RequestInit } from 'undici-types'

import { getHttpsUndiciAgent } from '../services/http.js'
import log from '../services/log.js'
import { isRegistryError } from '../types/RegistryError.js'
import config from '../utils/config.js'
import { InternalError, RegistryError } from '../utils/error.js'
import {
  isDoesLayerExistResponse,
  isGetImageTagManifestResponse,
  isInitialiseUploadObjectResponse,
  isListImageTagResponse,
  isListModelReposResponse,
  isPutManifestResponse,
  isRegistryErrorResponse,
  isUploadLayerMonolithicResponse,
} from './registryResponses.js'

interface RepoRef {
  namespace: string
  image: string
}

const registry = config.registry.connection.internal

const agent = getHttpsUndiciAgent({
  connect: { rejectUnauthorized: !config.registry.connection.insecure },
})

interface RegistryRequestResult<TBody = unknown> {
  headers: Record<string, string>
  body?: TBody
  stream?: ReadableStream | Readable
  abort: () => void
  status: number
  statusText: string
  url: string
}

async function registryRequest(
  token: string,
  endpoint: string,
  returnRawBody: boolean = false,
  extraFetchOptions: Partial<Omit<RequestInit, 'headers' | 'dispatcher' | 'signal'>> = {},
  extraHeaders: HeadersInit = {},
  pagination?: { traverseLinks?: boolean; start?: string; linkFilter?: string },
): Promise<RegistryRequestResult> {
  const controller = new AbortController()

  const allRepositories: string[] = []
  let paginateParameter = pagination?.start ? pagination.start : ''
  const linkFilter = pagination?.linkFilter ? pagination.linkFilter : ''
  let link: string | null
  let contentType: string
  let res: Response
  let body: any
  let stream: ReadableStream | Readable | undefined

  do {
    const url = `${registry}/v2/${endpoint}${paginateParameter}`
    log.debug({ url }, 'Making request to the registry.')
    try {
      // Note that this `fetch` is from `Node` and not `node-fetch` unlike other places in the codebase.
      // This is because `node-fetch` was incorrectly closing the stream received from `tar` for some (but not all) entries which meant that not all of the streamed data was sent to the registry
      res = await fetch(url, {
        headers: {
          Authorization: `Bearer ${token}`,
          ...extraHeaders,
        },
        dispatcher: agent,
        signal: controller.signal,
        ...extraFetchOptions,
      })
    } catch (err) {
      throw InternalError('Unable to communicate with the registry.', { err })
    }

    link = res.headers.get('link') || ''
    contentType = res.headers.get('content-type') || ''

    log.debug(Object.fromEntries(res.headers), 'Headers received from the registry.')

    const linkQueryIndex = link.indexOf('?')
    const linkEndIndex = link.lastIndexOf('>')
    if (link && linkQueryIndex !== -1 && linkEndIndex > linkQueryIndex) {
      paginateParameter = link.substring(linkQueryIndex, linkEndIndex)
    }

    if (returnRawBody) {
      stream = res.body as any
    } else if (contentType.endsWith('json')) {
      // e.g. 'application/json', 'application/vnd.docker.distribution.manifest.v2+json'
      try {
        body = await res.json()
      } catch (err) {
        throw InternalError('Unable to parse response body JSON.', { err })
      }
    } else {
      try {
        body = await res.text()
      } catch (err) {
        throw InternalError('Unable to read non-JSON response body.', { err })
      }
    }

    if (!res.ok) {
      const context = {
        url: res.url,
        status: res.status,
        statusText: res.statusText,
      }

      if (!body && contentType.includes('application/json')) {
        // try to get the json if there's an error, even if we wanted the raw body
        body = await res.json().catch(() => undefined)
      }

      if (isRegistryErrorResponse(body)) {
        throw RegistryError(body, context)
      } else {
        throw InternalError('Unrecognised response returned by the registry.', {
          ...context,
          body: JSON.stringify(body),
        })
      }
    }

    if (body?.repositories) {
      allRepositories.push(...body.repositories)
    }
  } while (pagination?.traverseLinks && link && link.includes(linkFilter))

  if (allRepositories.length) {
    body = {
      repositories: allRepositories,
    }
  }

  return {
    headers: res.headers ? Object.fromEntries(res.headers) : {},
    body: returnRawBody ? undefined : body,
    stream: returnRawBody ? stream : undefined,
    abort: () => controller.abort(),
    status: res.status,
    statusText: res.statusText,
    url: res.url,
  }
}

export async function listModelRepos(token: string, modelId: string) {
  const { body } = await registryRequest(token, '_catalog', undefined, undefined, undefined, {
    traverseLinks: true,
    start: `?n=100&last=${modelId}`,
    linkFilter: modelId,
  })
  if (!isListModelReposResponse(body)) {
    throw InternalError('Unrecognised response body when listing model repositories.', { body })
  }

  const filteredRepos = body.repositories.filter((repo) => repo.startsWith(`${modelId}/`))
  return filteredRepos
}

export async function listImageTags(token: string, imageRef: RepoRef) {
  const repo = `${imageRef.namespace}/${imageRef.image}`

  let body: unknown
  try {
    ;({ body } = await registryRequest(token, `${repo}/tags/list`, undefined, undefined, undefined, {
      traverseLinks: true,
    }))
  } catch (error) {
    if (isRegistryError(error) && error.errors.length === 1 && error.errors.at(0)?.code === 'NAME_UNKNOWN') {
      return []
    }
    throw error
  }

  if (!isListImageTagResponse(body)) {
    throw InternalError('Unrecognised response body when listing image tags.', { body })
  }
  return body.tags
}

export async function getImageTagManifest(token: string, imageRef: RepoRef, imageTag: string) {
  // TODO: handle `Accept: 'application/vnd.docker.distribution.manifest.list.v2+json'` type for multi-platform images
  const { body } = await registryRequest(
    token,
    `${imageRef.namespace}/${imageRef.image}/manifests/${imageTag}`,
    undefined,
    undefined,
    {
      Accept: 'application/vnd.docker.distribution.manifest.v2+json',
    },
  )
  if (!isGetImageTagManifestResponse(body)) {
    throw InternalError('Unrecognised response body when getting image tag manifest.', {
      body,
      namespace: imageRef.namespace,
      image: imageRef.image,
      imageTag,
    })
  }
  return body
}

export async function getRegistryLayerStream(
  token: string,
  imageRef: RepoRef,
  layerDigest: string,
): Promise<{ stream: Readable | ReadableStream; abort: () => void }> {
  const { stream, abort } = await registryRequest(
    token,
    `${imageRef.namespace}/${imageRef.image}/blobs/${layerDigest}`,
    true,
    undefined,
    {
      Accept: 'application/vnd.docker.distribution.manifest.v2+json',
    },
  )

  if (!stream || !(stream instanceof ReadableStream || stream instanceof Readable)) {
    abort()
    throw InternalError('Unrecognised response stream when getting image layer blob.', {
      stream,
      namespace: imageRef.namespace,
      image: imageRef.image,
      layerDigest,
    })
  }

  return { stream, abort }
}

export async function doesLayerExist(token: string, imageRef: RepoRef, digest: string) {
  try {
    const { headers } = await registryRequest(token, `${imageRef.namespace}/${imageRef.image}/blobs/${digest}`, true, {
      method: 'HEAD',
    })

    if (!isDoesLayerExistResponse(headers)) {
      throw InternalError('Unrecognised response headers when heading image layer.', {
        headers,
        namespace: imageRef.namespace,
        image: imageRef.image,
        digest,
      })
    }

    return true
  } catch (error) {
    if (typeof error === 'object' && error !== null && error['context']['status'] === 404) {
      // 404 response indicates that the layer does not exist
      return false
    } else {
      throw error
    }
  }
}

export async function initialiseUpload(token: string, imageRef: RepoRef) {
  const { headers } = await registryRequest(token, `${imageRef.namespace}/${imageRef.image}/blobs/uploads/`, true, {
    method: 'POST',
  })

  if (!isInitialiseUploadObjectResponse(headers)) {
    throw InternalError('Unrecognised response headers when posting initialise image upload.', {
      headers,
      namespace: imageRef.namespace,
      image: imageRef.image,
    })
  }

  return headers
}

export async function putManifest(
  token: string,
  imageRef: RepoRef,
  imageTag: string,
  manifest: BodyInit,
  contentType: string,
) {
  const { headers } = await registryRequest(
    token,
    `${imageRef.namespace}/${imageRef.image}/manifests/${imageTag}`,
    true,
    {
      method: 'PUT',
      body: manifest,
    },
    { 'Content-Type': contentType, name: `${imageRef.namespace}/${imageRef.image}`, reference: imageTag },
  )

  if (!isPutManifestResponse(headers)) {
    throw InternalError('Unrecognised response headers when putting image manifest.', {
      headers,
      namespace: imageRef.namespace,
      image: imageRef.image,
    })
  }

  return headers
}

export async function uploadLayerMonolithic(
  token: string,
  uploadURL: string,
  digest: string,
  blob: Readable | ReadableStream,
  size: string,
) {
  const { headers, abort } = await registryRequest(
    token,
    `${uploadURL}&digest=${digest}`.replace(/^(\/v2\/)/, ''),
    true,
    {
      method: 'PUT',
      body: blob,
      duplex: 'half',
      window: null,
      redirect: 'error',
    },
    {
      'content-length': size,
      'content-type': 'application/octet-stream',
    },
  )

  if (!isUploadLayerMonolithicResponse(headers)) {
    abort()
    throw InternalError('Unrecognised response headers when putting image blob.', {
      headers,
      uploadURL,
      digest,
      size,
    })
  }

  return headers
}
