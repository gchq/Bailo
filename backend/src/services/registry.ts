import { getImageTagManifest, getRegistryLayerStream, listImageTags, listModelRepos } from '../clients/registry.js'
import authorisation from '../connectors/authorisation/index.js'
import { UserInterface } from '../models/User.js'
import { getAccessToken } from '../routes/v1/registryAuth.js'
import { Forbidden } from '../utils/error.js'
import { getModelById } from './model.js'

export async function listModelImages(user: UserInterface, modelId: string) {
  const model = await getModelById(user, modelId)

  const auth = await authorisation.image(user, model, {
    type: 'repository',
    name: modelId,
    actions: ['list'],
  })
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, modelId })
  }

  const registryToken = await getAccessToken({ dn: user.dn }, [
    { type: 'registry', class: '', name: 'catalog', actions: ['*'] },
  ])
  const repos = await listModelRepos(registryToken, modelId)
  const versions = await Promise.all(
    repos.map(async (repo) => {
      const [namespace, image] = repo.split(/\/(.*)/s)
      const repositoryToken = await getAccessToken({ dn: user.dn }, [
        { type: 'repository', class: '', name: repo, actions: ['pull'] },
      ])
      return { repository: namespace, name: image, tags: await listImageTags(repositoryToken, { namespace, image }) }
    }),
  )

  return versions
}

export async function listImageTagLayers(user: UserInterface, modelId: string, imageName: string, imageTag: string) {
  const model = await getModelById(user, modelId)

  const auth = await authorisation.image(user, model, {
    type: 'repository',
    name: modelId,
    actions: ['pull'],
  })
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, modelId })
  }

  const repositoryToken = await getAccessToken({ dn: user.dn }, [
    { type: 'repository', class: '', name: `${modelId}/${imageName}`, actions: ['*'] },
  ])

  // get which layers exist for the model
  const manifest = await getImageTagManifest(repositoryToken, { namespace: modelId, image: imageName }, imageTag)

  return manifest.layers
}

export async function getImageBlob(user: UserInterface, modelId: string, imageName: string, digest: string) {
  const model = await getModelById(user, modelId)

  const auth = await authorisation.image(user, model, {
    type: 'repository',
    name: modelId,
    actions: ['pull'],
  })
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, modelId })
  }

  const repositoryToken = await getAccessToken({ dn: user.dn }, [
    { type: 'repository', class: '', name: `${modelId}/${imageName}`, actions: ['*'] },
  ])
  const responseBody = await getRegistryLayerStream(repositoryToken, { namespace: modelId, image: imageName }, digest)

  return responseBody
}
