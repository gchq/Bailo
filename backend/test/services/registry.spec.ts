import { describe, expect, test } from 'vitest'

import { joinDistributionPackageName, splitDistributionPackageName } from '../../src/services/registry.js'

describe('services > registry', () => {
  test('splitDistributionPackageName > success', () => {
    expect(splitDistributionPackageName('name:tag')).toMatchSnapshot()
    expect(splitDistributionPackageName('registry:3.0.0')).toMatchSnapshot()
    expect(splitDistributionPackageName('alpine:latest')).toMatchSnapshot()

    expect(splitDistributionPackageName('localhost:8080/name:tag')).toMatchSnapshot()
    expect(splitDistributionPackageName('clamav/clamav:1.4.2_base')).toMatchSnapshot()
    expect(splitDistributionPackageName('bitnamilegacy/minio:2025.4.22')).toMatchSnapshot()
    expect(splitDistributionPackageName('tarampampam/webhook-tester:latest')).toMatchSnapshot()
    expect(splitDistributionPackageName('marlonb/mailcrab:v1.5.0')).toMatchSnapshot()
    expect(splitDistributionPackageName('localhost:8080/export-4lvt8w/alpine:latest')).toMatchSnapshot()
    expect(
      splitDistributionPackageName('localhost:8080/exportfrom-v4yzsn/alpine/test/foo/bar:latest'),
    ).toMatchSnapshot()
    expect(splitDistributionPackageName('nginxinc/nginx-unprivileged:1.25.4-alpine3.18')).toMatchSnapshot()

    expect(splitDistributionPackageName('name@digest:0123456789abcdef0123456789abcdef')).toMatchSnapshot()
    expect(
      splitDistributionPackageName('registry@sha256:1fc7de654f2ac1247f0b67e8a459e273b0993be7d2beda1f3f56fbf1001ed3e7'),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName('alpine@sha256:a8560b36e8b8210634f77d9f7f9efd7ffa463e380b75e2e74aff4511df3ef88c'),
    ).toMatchSnapshot()

    expect(
      splitDistributionPackageName('localhost:8080/name@digest:0123456789abcdef0123456789abcdef'),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'clamav/clamav@sha256:e7d108f30ea8f16935dbd12e4b58665f1bc148ce3dd59028cf04088330216910',
      ),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'bitnamilegacy/minio@sha256:d7cd0e172c4cc0870f4bdc3142018e2a37be9acf04d68f386600daad427e0cab',
      ),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'tarampampam/webhook-tester@sha256:958a70683cbfdc8b150207b3f3732d0087df1c1a260e8b2f9cf0ec77dbedead3',
      ),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'marlonb/mailcrab@sha256:217db02005fbf51263789d63bfa63011011004685932d330ef8cefcfc054e8da',
      ),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'localhost:8080/export-4lvt8w/alpine@sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4',
      ),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'localhost:8080/exportfrom-v4yzsn/alpine/test/foo/bar@sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4',
      ),
    ).toMatchSnapshot()
    expect(
      splitDistributionPackageName(
        'nginxinc/nginx-unprivileged@sha256:7b4316677e4015a53d326e657915340128d9fd506f826f676d9f169c0c8557f6',
      ),
    ).toMatchSnapshot()
  })

  test('splitDistributionPackageName > error', () => {
    expect(() => splitDistributionPackageName('bad-name')).toThrowError('Could not parse Distribution Package Name.')
    expect(() => splitDistributionPackageName('foo:bar:baz')).toThrowError('Could not parse Distribution Package Name.')
    expect(() => splitDistributionPackageName('')).toThrowError('Could not parse Distribution Package Name.')
    expect(() => splitDistributionPackageName('bad-name@:0123456789abcdef0123456789abcdef')).toThrowError(
      'Could not parse Distribution Package Name.',
    )
    expect(() => splitDistributionPackageName('bad-name@sha256:0')).toThrowError(
      'Could not parse Distribution Package Name.',
    )
  })

  test('joinDistributionPackageName > success', () => {
    expect(joinDistributionPackageName({ path: 'name', tag: 'tag' })).toMatchSnapshot()
    expect(joinDistributionPackageName({ path: 'registry', tag: '3.0.0' })).toMatchSnapshot()
    expect(joinDistributionPackageName({ path: 'alpine', tag: 'latest' })).toMatchSnapshot()

    expect(
      joinDistributionPackageName({ path: 'name', digest: 'digest:0123456789abcdef0123456789abcdef' }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        path: 'registry',
        digest: 'sha256:1fc7de654f2ac1247f0b67e8a459e273b0993be7d2beda1f3f56fbf1001ed3e7',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        path: 'alpine',
        digest: 'sha256:a8560b36e8b8210634f77d9f7f9efd7ffa463e380b75e2e74aff4511df3ef88c',
      }),
    ).toMatchSnapshot()

    expect(joinDistributionPackageName({ domain: 'localhost:8080', path: 'name', tag: 'tag' })).toMatchSnapshot()
    expect(joinDistributionPackageName({ domain: 'clamav', path: 'clamav', tag: '1.4.2_base' })).toMatchSnapshot()
    expect(joinDistributionPackageName({ domain: 'bitnamilegacy', path: 'minio', tag: '2025.4.22' })).toMatchSnapshot()
    expect(
      joinDistributionPackageName({ domain: 'tarampampam', path: 'webhook-tester', tag: 'latest' }),
    ).toMatchSnapshot()
    expect(joinDistributionPackageName({ domain: 'marlonb', path: 'mailcrab', tag: 'v1.5.0' })).toMatchSnapshot()
    expect(
      joinDistributionPackageName({ domain: 'localhost:8080', path: 'export-4lvt8w/alpine', tag: 'latest' }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'localhost:8080',
        path: 'exportfrom-v4yzsn/alpine/test/foo/bar',
        tag: 'latest',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({ path: 'nginxinc/nginx-unprivileged', tag: '1.25.4-alpine3.18' }),
    ).toMatchSnapshot()

    expect(
      joinDistributionPackageName({
        domain: 'localhost:8080',
        path: 'name',
        digest: 'sha256:0123456789abcdef0123456789abcdef',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'clamav',
        path: 'clamav',
        digest: 'sha256:e7d108f30ea8f16935dbd12e4b58665f1bc148ce3dd59028cf04088330216910',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'bitnamilegacy',
        path: 'minio',
        digest: 'sha256:d7cd0e172c4cc0870f4bdc3142018e2a37be9acf04d68f386600daad427e0cab',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'tarampampam',
        path: 'webhook-tester',
        digest: 'sha256:958a70683cbfdc8b150207b3f3732d0087df1c1a260e8b2f9cf0ec77dbedead3',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'marlonb',
        path: 'mailcrab',
        digest: 'sha256:217db02005fbf51263789d63bfa63011011004685932d330ef8cefcfc054e8da',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'localhost:8080',
        path: 'export-4lvt8w/alpine',
        digest: 'sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        domain: 'localhost:8080',
        path: 'exportfrom-v4yzsn/alpine/test/foo/bar',
        digest: 'sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4',
      }),
    ).toMatchSnapshot()
    expect(
      joinDistributionPackageName({
        path: 'nginxinc/nginx-unprivileged',
        digest: 'sha256:7b4316677e4015a53d326e657915340128d9fd506f826f676d9f169c0c8557f6',
      }),
    ).toMatchSnapshot()
  })

  test('joinDistributionPackageName > error', () => {
    expect(() => joinDistributionPackageName({ path: 'bad-name', tag: '' })).toThrowError(
      'Could not join Distribution Package Name.',
    )
    expect(() => joinDistributionPackageName({ path: '', tag: 'foo:bar:baz' })).toThrowError(
      'Could not join Distribution Package Name.',
    )
    expect(() => joinDistributionPackageName({ path: '', tag: '' })).toThrowError(
      'Could not join Distribution Package Name.',
    )
    expect(() => joinDistributionPackageName({ path: '', digest: ':0123456789abcdef0123456789abcdef' })).toThrowError(
      'Could not join Distribution Package Name.',
    )
    expect(() => joinDistributionPackageName({ domain: 'bad-name', path: '', digest: 'sha256:0' })).toThrowError(
      'Could not join Distribution Package Name.',
    )
  })

  test('joinDistributionPackageName -> splitDistributionPackageName > success', () => {
    const testObjects = [
      { path: 'name', tag: 'tag' },
      { path: 'registry', tag: '3.0.0' },
      { path: 'alpine', tag: 'latest' },
      { path: 'nginxinc/nginx-unprivileged', tag: '1.25.4-alpine3.18' },
      { path: 'name', digest: 'digest:0123456789abcdef0123456789abcdef' },
      {
        path: 'registry',
        digest: 'sha256:1fc7de654f2ac1247f0b67e8a459e273b0993be7d2beda1f3f56fbf1001ed3e7',
      },
      {
        path: 'alpine',
        digest: 'sha256:a8560b36e8b8210634f77d9f7f9efd7ffa463e380b75e2e74aff4511df3ef88c',
      },
      {
        path: 'nginxinc/nginx-unprivileged',
        digest: 'sha256:7b4316677e4015a53d326e657915340128d9fd506f826f676d9f169c0c8557f6',
      },
      {
        domain: 'localhost:8080',
        path: 'export-4lvt8w/alpine',
        digest: 'sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4',
      },
      {
        path: 'node',
        tag: '24.4.1-alpine',
      },
      {
        domain: 'localhost:8080',
        path: 's-3l1kaq/tensorflow/tensorflow',
        tag: 'latest-gpu-jupyter',
      },
      {
        path: 'tensorflow/tensorflow',
        tag: 'latest-gpu-jupyter',
      },
    ]

    for (const testObject of testObjects) {
      expect(splitDistributionPackageName(joinDistributionPackageName(testObject))).toStrictEqual(testObject)
    }
  })

  test('splitDistributionPackageName -> joinDistributionPackageName > success', () => {
    const testStrings = [
      'name:tag',
      'registry:3.0.0',
      'alpine:latest',
      'nginxinc/nginx-unprivileged:1.25.4-alpine3.18',
      'name@digest:0123456789abcdef0123456789abcdef',
      'registry@sha256:1fc7de654f2ac1247f0b67e8a459e273b0993be7d2beda1f3f56fbf1001ed3e7',
      'alpine@sha256:a8560b36e8b8210634f77d9f7f9efd7ffa463e380b75e2e74aff4511df3ef88c',
      'nginxinc/nginx-unprivileged@sha256:7b4316677e4015a53d326e657915340128d9fd506f826f676d9f169c0c8557f6',
      'localhost:8080/export-4lvt8w/alpine/sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4',
      'node:24.4.1-alpine',
      'localhost:8080/s-3l1kaq/tensorflow/tensorflow:latest-gpu-jupyter',
      'tensorflow/tensorflow:latest-gpu-jupyter',
    ]

    for (const testString of testStrings) {
      expect(joinDistributionPackageName(splitDistributionPackageName(testString))).toStrictEqual(testString)
    }
  })
})
