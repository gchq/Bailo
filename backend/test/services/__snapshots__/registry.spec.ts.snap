// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > registry > joinDistributionPackageName > success 1`] = `"name:tag"`;

exports[`services > registry > joinDistributionPackageName > success 2`] = `"registry:3.0.0"`;

exports[`services > registry > joinDistributionPackageName > success 3`] = `"alpine:latest"`;

exports[`services > registry > joinDistributionPackageName > success 4`] = `"name@digest:0123456789abcdef0123456789abcdef"`;

exports[`services > registry > joinDistributionPackageName > success 5`] = `"registry@sha256:1fc7de654f2ac1247f0b67e8a459e273b0993be7d2beda1f3f56fbf1001ed3e7"`;

exports[`services > registry > joinDistributionPackageName > success 6`] = `"alpine@sha256:a8560b36e8b8210634f77d9f7f9efd7ffa463e380b75e2e74aff4511df3ef88c"`;

exports[`services > registry > joinDistributionPackageName > success 7`] = `"localhost:8080/name:tag"`;

exports[`services > registry > joinDistributionPackageName > success 8`] = `"clamav/clamav:1.4.2_base"`;

exports[`services > registry > joinDistributionPackageName > success 9`] = `"bitnamilegacy/minio:2025.4.22"`;

exports[`services > registry > joinDistributionPackageName > success 10`] = `"tarampampam/webhook-tester:latest"`;

exports[`services > registry > joinDistributionPackageName > success 11`] = `"marlonb/mailcrab:v1.5.0"`;

exports[`services > registry > joinDistributionPackageName > success 12`] = `"localhost:8080/export-4lvt8w/alpine:latest"`;

exports[`services > registry > joinDistributionPackageName > success 13`] = `"localhost:8080/exportfrom-v4yzsn/alpine/test/foo/bar:latest"`;

exports[`services > registry > joinDistributionPackageName > success 14`] = `"nginxinc/nginx-unprivileged:1.25.4-alpine3.18"`;

exports[`services > registry > joinDistributionPackageName > success 15`] = `"localhost:8080/name@sha256:0123456789abcdef0123456789abcdef"`;

exports[`services > registry > joinDistributionPackageName > success 16`] = `"clamav/clamav@sha256:e7d108f30ea8f16935dbd12e4b58665f1bc148ce3dd59028cf04088330216910"`;

exports[`services > registry > joinDistributionPackageName > success 17`] = `"bitnamilegacy/minio@sha256:d7cd0e172c4cc0870f4bdc3142018e2a37be9acf04d68f386600daad427e0cab"`;

exports[`services > registry > joinDistributionPackageName > success 18`] = `"tarampampam/webhook-tester@sha256:958a70683cbfdc8b150207b3f3732d0087df1c1a260e8b2f9cf0ec77dbedead3"`;

exports[`services > registry > joinDistributionPackageName > success 19`] = `"marlonb/mailcrab@sha256:217db02005fbf51263789d63bfa63011011004685932d330ef8cefcfc054e8da"`;

exports[`services > registry > joinDistributionPackageName > success 20`] = `"localhost:8080/export-4lvt8w/alpine@sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4"`;

exports[`services > registry > joinDistributionPackageName > success 21`] = `"localhost:8080/exportfrom-v4yzsn/alpine/test/foo/bar@sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4"`;

exports[`services > registry > joinDistributionPackageName > success 22`] = `"nginxinc/nginx-unprivileged@sha256:7b4316677e4015a53d326e657915340128d9fd506f826f676d9f169c0c8557f6"`;

exports[`services > registry > splitDistributionPackageName > success 1`] = `
{
  "domain": "",
  "path": "name",
  "tag": "tag",
}
`;

exports[`services > registry > splitDistributionPackageName > success 2`] = `
{
  "domain": "",
  "path": "registry",
  "tag": "3.0.0",
}
`;

exports[`services > registry > splitDistributionPackageName > success 3`] = `
{
  "domain": "",
  "path": "alpine",
  "tag": "latest",
}
`;

exports[`services > registry > splitDistributionPackageName > success 4`] = `
{
  "domain": "localhost:8080",
  "path": "name",
  "tag": "tag",
}
`;

exports[`services > registry > splitDistributionPackageName > success 5`] = `
{
  "domain": "clamav",
  "path": "clamav",
  "tag": "1.4.2_base",
}
`;

exports[`services > registry > splitDistributionPackageName > success 6`] = `
{
  "domain": "bitnamilegacy",
  "path": "minio",
  "tag": "2025.4.22",
}
`;

exports[`services > registry > splitDistributionPackageName > success 7`] = `
{
  "domain": "tarampampam",
  "path": "webhook-tester",
  "tag": "latest",
}
`;

exports[`services > registry > splitDistributionPackageName > success 8`] = `
{
  "domain": "marlonb",
  "path": "mailcrab",
  "tag": "v1.5.0",
}
`;

exports[`services > registry > splitDistributionPackageName > success 9`] = `
{
  "domain": "localhost:8080",
  "path": "export-4lvt8w/alpine",
  "tag": "latest",
}
`;

exports[`services > registry > splitDistributionPackageName > success 10`] = `
{
  "domain": "localhost:8080",
  "path": "exportfrom-v4yzsn/alpine/test/foo/bar",
  "tag": "latest",
}
`;

exports[`services > registry > splitDistributionPackageName > success 11`] = `
{
  "domain": "nginxinc",
  "path": "nginx-unprivileged",
  "tag": "1.25.4-alpine3.18",
}
`;

exports[`services > registry > splitDistributionPackageName > success 12`] = `
{
  "digest": "digest:0123456789abcdef0123456789abcdef",
  "domain": "",
  "path": "name",
}
`;

exports[`services > registry > splitDistributionPackageName > success 13`] = `
{
  "digest": "sha256:1fc7de654f2ac1247f0b67e8a459e273b0993be7d2beda1f3f56fbf1001ed3e7",
  "domain": "",
  "path": "registry",
}
`;

exports[`services > registry > splitDistributionPackageName > success 14`] = `
{
  "digest": "sha256:a8560b36e8b8210634f77d9f7f9efd7ffa463e380b75e2e74aff4511df3ef88c",
  "domain": "",
  "path": "alpine",
}
`;

exports[`services > registry > splitDistributionPackageName > success 15`] = `
{
  "digest": "digest:0123456789abcdef0123456789abcdef",
  "domain": "localhost:8080",
  "path": "name",
}
`;

exports[`services > registry > splitDistributionPackageName > success 16`] = `
{
  "digest": "sha256:e7d108f30ea8f16935dbd12e4b58665f1bc148ce3dd59028cf04088330216910",
  "domain": "clamav",
  "path": "clamav",
}
`;

exports[`services > registry > splitDistributionPackageName > success 17`] = `
{
  "digest": "sha256:d7cd0e172c4cc0870f4bdc3142018e2a37be9acf04d68f386600daad427e0cab",
  "domain": "bitnamilegacy",
  "path": "minio",
}
`;

exports[`services > registry > splitDistributionPackageName > success 18`] = `
{
  "digest": "sha256:958a70683cbfdc8b150207b3f3732d0087df1c1a260e8b2f9cf0ec77dbedead3",
  "domain": "tarampampam",
  "path": "webhook-tester",
}
`;

exports[`services > registry > splitDistributionPackageName > success 19`] = `
{
  "digest": "sha256:217db02005fbf51263789d63bfa63011011004685932d330ef8cefcfc054e8da",
  "domain": "marlonb",
  "path": "mailcrab",
}
`;

exports[`services > registry > splitDistributionPackageName > success 20`] = `
{
  "digest": "sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4",
  "domain": "localhost:8080",
  "path": "export-4lvt8w/alpine",
}
`;

exports[`services > registry > splitDistributionPackageName > success 21`] = `
{
  "digest": "sha256:ec1b05d1eac264d9204a57f4ad9d4dc35e9e756e9fedaea0674aefc7edb1d6a4",
  "domain": "localhost:8080",
  "path": "exportfrom-v4yzsn/alpine/test/foo/bar",
}
`;

exports[`services > registry > splitDistributionPackageName > success 22`] = `
{
  "digest": "sha256:7b4316677e4015a53d326e657915340128d9fd506f826f676d9f169c0c8557f6",
  "domain": "nginxinc",
  "path": "nginx-unprivileged",
}
`;
